### PR TITLE
Adding support for multiple Protocols.

### PR DESCRIPTION
**Pull request by cdunn2001**
Not yet ready to pull, but try it.
`
doozerd &
GetWidgets/rand-agent/rand-agent -protocol=json &
GetWidgets/initiator/rand-init/rand-init
GetWidgets/initiator/rand-init/rand-init
GetWidgets/initiator/rand-init/rand-init
`
Works for me!
`
GetWidgets/$ initiators/rand-init/rand-init
Request: 3
Reponse: dqo
Done.
GetWidgets/$ initiators/rand-init/rand-init
Request: 9
Reponse: rnobhiyyd
Done.
GetWidgets/$ initiators/rand-init/rand-init
Request: 10
Reponse: nfxunamqlc
Done.
`
## Notes
- **rand-init**: Initiator to call MyService.RandString with random arg.
- **rand-agent**: Agent to return a RandomWord of given length.
- New flag: `-protocol`
- This still needs lots of refactoring, debugging, and better logging.

### CODE REVIEW

File: examples/GetWidgets/initiators/web/main.go [hunk 1]
```diff
@@ -99,7 +95,7 @@ func main() {
 	http.HandleFunc("/", homeHandler)
 	http.HandleFunc("/new", submitHandler)
 
-	rpc.HandleHTTP()
+	rpc.HandleHTTP() // I think we still need this here.
```
Comment (bketelsen) [inline examples/GetWidgets/initiators/web/main.go:98]: This can't go in sky lib?

**Comment by bketelsen** [comment]
Does this do auto-port assignment too?  This looks like almost exactly what we have been discussing.  Simple for the service author to implement.

File: examples/GetWidgets/initiators/web/main.go (hunk 1 above)
Comment (cdunn2001) [inline examples/GetWidgets/initiators/web/main.go:98]: It _is_ in skylib, in `Node::Serve()`, which is called by Agents, not by Initiators. (It is called by `Start()` for each registered Server.)

All it does is call `http.Handle()`, similar to the 2 previous lines, but for `rpcPath` and `rpcDebugPath`. It's safe to call it twice, though that won't happen in this case.

But I'm not sure it's really called for here, since the web Initiator is probably not expected to serve RPC requests from the internet.

**Comment by cdunn2001** [comment]
Auto-port assignment is not there yet. That's a nice, well-defined task to let another volunteer get involved. If there is an issue for it, you could offer that to somebody on skynet-dev. But we need to get this basic re-factoring agreed, finalized, and merged first, I think.

Yes, it's basically what we've been discussion (sans **reddis**), but anything can be changed. I am a big fan of proto-typing, so that discussions become more concrete. I am not yet sure whether the original Agents (router, watcher, etc.) have been properly migrated either.

File: examples/GetWidgets/initiators/web/main.go (hunk 1 above)
Comment (bketelsen) [inline examples/GetWidgets/initiators/web/main.go:98]: aahhh I see.. it's hard to follow these on my iPhone (on vacation).  The initiators have rpc calls for service.ping requests, which is what the reaper uses for keepalives.

Comment (cdunn2001) [inline examples/GetWidgets/initiators/web/main.go:98]: First, a change in terminology, which seems to be what [the the SOA world is using](http://www.via-nova-architectura.org/files/Reviewed/Cace.pdf):

```
class FooInterface {
    int FooOperation1(req, resp);
    int FooOperation2(req, resp);
};

class CommonInterface {
    int Ping(req, resp);
};
```

The Interfaces define groups of Operations. (A Contract would include the Interface plus any pre and post conditions, and possibly the transport protocol.) With these Interfaces, a Node would have at least one server, maybe more. A Service is then the actual provision of those operations by a given Server.

I'll re-adjust my code to that and re-start the Pull request, in a moment.

---

For the Heartbeat, I would have every Node auto-register a Ping Operation amongst the CommonInterface. In that case, every Node has at least one server.

Then, instead of walking through all Services and calling Ping as if each were a separate process (recall that an Agent can offer multiple Services, each of which may provide multiple Operations), the monitor could grab just the CommonInterface Services and call Ping on each of them.

There is a question of whether to offer both Interfaces on the same port, by the same Server. That would be cheaper (in terms of numbers of Ports) but a long-running Operation would block the Ping request.

Comment (cdunn2001) [inline examples/GetWidgets/initiators/web/main.go:98]: Actually, [this](http://www.soa-in-practice.com/soa-glossary.html) might be more recent terminology. They are using "Provider" where I used "Agent"; they use "Agent" where I used "Node", and they use "Model" where I used "Interface" (or earlier, "Provision"). I'll switch to those.

Comment (bketelsen) [inline examples/GetWidgets/initiators/web/main.go:98]: I like these better.
